### PR TITLE
Fix concurrent user activation throttler

### PR DESCRIPTION
**Pull request by markusthoemmes**
- ConsulClient failed silently and with the wrong exception
- A failure in one call (not even related to the other calls) tore down the entire update logic

### CODE REVIEW

**Comment by markusthoemmes** [comment]
Note to self: Try to optimize the numbers of action invocations by stopping as soon as they get throttled.

File: core/controller/src/main/scala/whisk/core/loadBalancer/ActivationThrottle.scala [hunk 1]
```diff
@@ -120,22 +120,25 @@ class ActivationThrottle(config: WhiskConfig)(
...
-        for {
-            concurrencyLimits <- getConcurrencyLimits
+        // Fall back to the current version of the limits
+        val concurrency = getConcurrencyLimits.recover {
+            case _ => userActivationLimits
+        }.map { concurrencyLimits =>
+            userActivationLimits = concurrencyLimits
+        }
+
+        val activations = for {
```
Comment (psuter) [inline core/controller/src/main/scala/whisk/core/loadBalancer/ActivationThrottle.scala]: No fallback for this case?

Comment (markusthoemmes) [inline core/controller/src/main/scala/whisk/core/loadBalancer/ActivationThrottle.scala]: If one of both calls fail, the calculation would be invalid anyway.

Your comment made me notice though, that above "fallback" is unnecessary. Removing.

**Comment by markusthoemmes** [comment]
@perryibm please take a look at the test specifically

**Comment by perryibm** [comment]
3 suggestions:

1) Move the throttle window `1.minute` to a constant and group it with the 60.   The time window of 1 minute and 60 firings are whisk limit properties and these can be adjusted as needed if we start changing the way we throttle.
2) The parameter maximumInvoke of 120 is really a testing parameter and perhaps shoud be restated as 2*maximumFirings to show that we decide to go at twice the throttle rate to trigger it.
3) Once something is throttled, it would be great if you could stop the Future's that have not been fired from occurring.  One way would be through a condition variable but I leave it to you to do it or not.  It would be good to know how long this test takes on average to complete.  In a perfect world, it would be 1 minute.  When one gets unlucky, it would be 2 minutes if we coded things well.  Since you've gone to the trouble if optimizing, can you indicate what you usually get?

**Comment by markusthoemmes** [comment]
Thanks @perry, i'll have a look at the constants.

To 3: Already did that, check the last commit. It stops as soon as a request is throttled.

**Comment by rabbah** [comment]
Can we lift the rate limits out of the throttle and parameterize this test accordingly --- If we change the limits upward, will this test catch the lack of throttling?

**Comment by markusthoemmes** [comment]
@perryibm in the tests take somewhere between 4 to 5 minutes.

**Comment by rabbah** [comment]
Can we move these bits from LoadBalancerToKafka to the RateThorttle?

```
                val userCount = activationThrottle.countForNamespace(subject)
                val userLimit = activationThrottle.limitForNamespace(subject)
                info(this, s"(DoS) current activation count for '$subject': $userCount (limit=$userLimit)")
                if (userCount > userLimit) {
                    info(this, s"(DoS) '$subject' maxed concurrent invocations")
                    Future.successful(throttleError)
```

File: tests/src/limits/ThrottleTests.scala [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+
+    /**
+     * Extracts the number of throttled results from a sequence of <code>RunResult</code>
+     *
+     * @param results the sequence of results
+     * @param message the message to determine the type of throttling
+     */
+    def throttledActivations(results: List[RunResult], message: String) = results.map { result =>
+        result.stderr
+    }.count(_.contains(message))
```
Comment (markusthoemmes) [inline tests/src/limits/ThrottleTests.scala]: Memo to self: Rewrite as `results.count(_.stderr.contains(message))`

File: tests/src/limits/ThrottleTests.scala [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+            }
+
+            // invokes per minute * 2 because the current minute could advance which resets the throttle
+            val results = List.fill(maximumInvokesPerMinute * 2 + 1)("testWord").par.map { payload =>
+                wsk.action.invoke(name, Map("payload" -> payload.toJson), expectedExitCode = DONTCARE_EXIT)
+            }
+            val afterInvokes = Instant.now
+
+            waitForActivations(results)
+            throttledActivations(results, rateMessage) should be > 0
```
Comment (rabbah) [inline tests/src/limits/ThrottleTests.scala]: why not check the exit code is "throttled"?

File: tests/src/limits/ThrottleTests.scala [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+
+    /**
+     * Extracts the number of throttled results from a sequence of <code>RunResult</code>
+     *
+     * @param results the sequence of results
+     * @param message the message to determine the type of throttling
+     */
+    def throttledActivations(results: ParSeq[RunResult], message: String) = results.map { result =>
+        result.stderr
+    }.count(_.contains(message))
```
Comment (rabbah) [inline tests/src/limits/ThrottleTests.scala]: contains message && exit code is "throttled" for completeness?

File: core/controller/src/main/scala/whisk/core/entitlement/Entitlement.scala [hunk 5]
```diff
@@ -90,6 +91,7 @@ protected[core] abstract class EntitlementService(config: WhiskConfig)(
     private var loadbalancerOverload: Option[Boolean] = None
     private val invokeRateThrottler = new RateThrottler(config.actionInvokePerMinuteLimit.toInt, config.actionInvokePerHourLimit.toInt)
     private val triggerRateThrottler = new RateThrottler(config.triggerFirePerMinuteLimit.toInt, config.triggerFirePerHourLimit.toInt)
+    private val concurrentInvokeThrottler = new ActivationThrottler(config.consulServer, config.actionInvokeConcurrentLimit.toLong)
```
Comment (rabbah) [inline core/controller/src/main/scala/whisk/core/entitlement/Entitlement.scala]: long vs int?

**Comment by markusthoemmes** [comment]
@rabbah moved the code into the entitlement package, factored in the "long vs int" comment. Squashed to what i think are the 3 orthogonal bits in this PR.

**Comment by markusthoemmes** [comment]
PG1 219 is good.

**Comment by rabbah** [comment]
Merged as https://github.com/openwhisk/openwhisk/compare/e5cf3c1806b8...2895a9626c84.

**Pull request closed without merging**